udev: Add id program for hidraw devices (e.g. security tokens)

FabianHenneke opened the pull request:
Add a hidraw_id program meant to be run for devices in the hidraw subsystem via
an IMPORT directive. The program parses the HID report descriptor and assigns
environment variables based on the declared usage pages and usages.

The currently only purpose of this program is to automatically label FIDO
security tokens with ID_SECURITY_TOKEN, which allows access from userspace.
So far, every new security token model had to be manually whitelisted.

This commit is accompanied by a test suite and a fuzzer target for the
descriptor parsing routine.

yuwata commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+        if (desclen <= 0)
+                goto out;
+
+        if (parse_hid_report_desc(desc, desclen, &desc_result) != 0) {
```
We usually returns negative errno on error, and please do not ignore the returned value.
```
r = parse_hid_report_desc();
if (r < 0) {
    log_debug_errno(r, "Failed to parse report descriptor at '%s': %m", path);
    goto out;
}
```

yuwata commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+
+        desclen = read(fd, desc, sizeof(desc));
+        if (desclen <= 0)
+                goto out;
```
```
if (desclen < 0)
    return log_debug_errno(errno, "...");
if (desclen == 0)
    return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "...");
```

yuwata commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+
+        fd = open(path, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+        if (fd == -1) {
+                log_debug("Failed to open report descriptor at: %s", path);
```
`log_debug_errno(errno, "...");`

poettering commented:
commit msg should say "Fixes: #11996"

evverx commented on src/fuzz/fuzz-hidraw-id-desc.options:
```diff
@@ -0,0 +1,2 @@
+[libfuzzer]
+max_len = 4096
```
Could you remove this file and add something like
```c
if (size > 4096)
    return 0;
```
to the fuzzer?

In the meantime, I've just opened to https://github.com/systemd/systemd/pull/13361 to get rid of all the "options" files we don't need any more.

FabianHenneke replied:
I added the `max_len` config mostly because I thought libFuzzer would set it to the size of the largest file in the corpus otherwise. The corpus currently consists of valid HID report descriptors which are rather small compared to their theoretical maximal length. Would you recommend to include larger samples in the corpus instead?

evverx replied:
On ClusterFuzz when `libFuzzer` is used as a fuzzing engine some random values between 1 and 10000 is picked and passed to the fuzzers, which kind of helps us to generate corpora with no seed corpus whatsoever :-) As far as I know it's called `random_max_len_strategy` there. I'm not sure what's going on Fuzzit (where *.options files aren't supported) though. So far I've just moved the corpora OSS-Fuzz has been accumulating for over a year to Fuzzit. @yevgenypats could you help us out? Would it make sense to pass `-max-len` to fuzzers on Fuzzit to avoid the default behavior of libFuzzer?

Anyway, usually we don't keep corpora in the repository (mostly because they are relatively large even when compressed). Instead we start with several files just to make sure that the basics are covered and eventually add testcases that have triggered any issues on ClusterFuzz, Fuzzit and so on. So I'd just leave the 2 files where they are and add a testcase crashing the fuzzer under UBsan. Hope it makes sense.

yevgenypats replied:
I think `-max-len 4096` is the default behaviour anyway even in case of a large test-case. I'll validate that now.

FabianHenneke replied:
> I think -max-len 4096 is the default behaviour anyway even in case of a large test-case. I'll validate that now.

Is it also the default in case of a small test case? The corpora consists of files with ~65 bytes, but we would like to generate samples up to 4096 bytes.

yevgenypats replied:
Per small files, yes, for sure. per large files yes that makes sense to pass `-max_len=4096`, wasn't aware of this but now tested and it is indeed uses the largest file as max_len or 4096

FabianHenneke replied:
Thanks for testing this! As a result, I will remove the `.options` file.

evverx commented on src/fuzz/fuzz-hidraw-id-desc.c:
```diff
@@ -0,0 +1,20 @@
+        if (!getenv("SYSTEMD_LOG_LEVEL"))
+                log_set_max_level(LOG_CRIT);
+
+        parse_hid_report_desc(data, size, &res);
```
Generally we use `(void)` if we ignore the result of a function to silence static analysers but in this particular case I think it would be better to process `res` mostly to help MSan to make sure it's initialized properly after `parse_hid_report_desc` is finished.

FabianHenneke replied:
Could you verify that I simulate this processing in the correct way? I am storing a value depending on everything in res in a volatile non-local variable, which I do in order to get around both compiler optimizations and warnings about unused variables.

evverx replied:
@FabianHenneke MSan isn't triggered when uninitialized memory is copied directly without using anywhere so the assignment is unlikely to work even though the variable doesn't seem to be compiled out. I think we should start using `__msan_check_mem_is_initialized` at some point but to be honest it's the rabbit hole I'd like to go down later. I think the original version with `(void)` before `parse_hid_report_desc` should be good enough for now.

evverx replied:
I went through our fuzzers and it turns out we almost never try to make sure that whatever we get after parsing data is properly initialized and most likely miss bugs that can easily 
 be caught by MSan. I'll add it to the fuzzing TODO list right after i386 support I started to play with in https://github.com/systemd/systemd/pull/12608.

I think it might be a good idea to just get back to `(void) parse_hid_report_desc()`. Sorry about the misdirection.

FabianHenneke replied:
No worries, thank you for helping out with the fuzzer results. I'm open to explore the use of `__msan_check_mem_is_initialized` once it reaches the top of that list.

evverx commented:
I'll go ahead and add the "ci-fails" label because Travis will fail until a new fuzz target is created on Fuzzit. Let me know when the PR is more or less ready to roll so that I can do that.

Other than that, the fuzzer seems to be failing under UBSan:
```sh
#7489	NEW    cov: 34 ft: 94 corp: 31/92b lim: 6 exec/s: 0 rss: 32Mb L: 6/6 MS: 2 InsertByte-InsertByte-
../src/udev/hidraw_id/hidraw_id_desc.c:47:51: runtime error: left shift of 242 by 24 places cannot be represented in type 'int'
    #0 0x464d22 in parse_hid_report_desc /home/vagrant/systemd/build/../src/udev/hidraw_id/hidraw_id_desc.c:47:51
    #1 0x46494f in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/fuzz/fuzz-hidraw-id-desc.c:17:9
    #2 0x41dc7d in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x41dc7d)
    #3 0x41ec05 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x41ec05)
    #4 0x4207d8 in fuzzer::Fuzzer::MutateAndTestOne() (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x4207d8)
    #5 0x42342f in fuzzer::Fuzzer::Loop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, fuzzer::fuzzer_allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x42342f)
    #6 0x412f77 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x412f77)
    #7 0x406bb6 in main (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x406bb6)
    #8 0x7f1d46dcf412 in __libc_start_main (/lib64/libc.so.6+0x24412)
    #9 0x406c0d in _start (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x406c0d)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/udev/hidraw_id/hidraw_id_desc.c:47:51 in
MS: 5 ChangeByte-ShuffleBytes-CopyPart-ChangeByte-CMP- DE: "\xff\xff\xff\xff"-; base unit: a42c6cf1de3abfdea9b95f34687cbbe92b9a7383
0xff,0xff,0xff,0xff,0xf2,
\xff\xff\xff\xff\xf2
artifact_prefix='./'; Test unit written to ./crash-5e5f6feb192657393363a126c004f3a161eb9e3e
Base64: //////I=
```
It can be reproduced by installing `clang` and running
```
SANITIZER=undefined ./tools/oss-fuzz.sh
UBSAN_OPTIONS=print_stacktrace=1:print_summary=1:halt_on_error=1 ./out/fuzz-hidraw-id-desc
```

poettering commented on rules/60-hidraw_id.rules:
```diff
@@ -0,0 +1,7 @@
+
+ACTION=="remove", GOTO="hidraw_id_end"
+
+SUBSYSTEM=="hidraw", IMPORT{program}="hidraw_id --udev"
```
hmm, isn't there supposed to be a rules addition that adds uaccess to this, too?

(btw, are we sure that giving unpriv clients hidraw access is generally considered safe? is the kernel device API for this designed to be usable by unpriv client?)

poettering replied:
@amluto I understand you wrote the original code this PR is based on, maybe you can comment on this: is the hidraw kernel subsystem designed to be safe for unpriv userspace to access it? i.e. if we now open access to such devices by default, is that a good and safe idea?

FabianHenneke replied:
> hmm, isn't there supposed to be a rules addition that adds uaccess to this, too?

That rule is already part of systemd, `uaccess` is added for `ID_SECURITY_TOKEN` by `src/login/70-uaccess.rules.m4`. The `ID_U2F_TOKEN` is there solely for full compatibility with the existing `u2f-hidraw-policy` package, which is part of e.g. Fedora.

> (btw, are we sure that giving unpriv clients hidraw access is generally considered safe? is the kernel device API for this designed to be usable by unpriv client?)

I am not knowledgeable enough about its internals to declare it safe, but I can say that (1) having raw access to (correctly designed) security tokens is safe since every meaningful action tests for user presence first and (2) unprivileged clients do already have access to whitelisted security tokens. I would also be interested in @amluto's opinion.

amluto replied:
> @amluto I understand you wrote the original code this PR is based on, maybe you can comment on this: is the hidraw kernel subsystem designed to be safe for unpriv userspace to access it? i.e. if we now open access to such devices by default, is that a good and safe idea?

As far as I know, this is the intent.  It's been used for exactly this purpose with U2F devices for years, so, if there's a problem, it's not systemd's fault :)

FabianHenneke commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+
+        desclen = read(fd, desc, sizeof(desc));
+        if (desclen <= 0)
+                goto out;
```
I decided to error out when accessing the device or its descriptor fails, but only log debug messages if something should be amiss with the descriptor itself. Do you think that's reasonable?

FabianHenneke commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,71 @@
+                        /* Long item; skip it. */
+                        if (i + 1 >= desclen)
+                                return -EINVAL;
+                        i += (desc[i+1] + 3); /* Can't overflow. */
```
I added comments explaining the relevant parts of the HID standard throughout.

FabianHenneke commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,71 @@
+                        value |= (desc[i + 1 + j] << 8*j);
+
+                if (type == HID_RPTDESC_TYPE_GLOBAL && tag == HID_RPTDESC_GLOBAL_ITEM_USAGE_PAGE)
+                        usage_page = value;
```
Very fishy indeed, thanks for catching this. Thankfully it was caught  by the fuzzer.
I rewrote most of the parsing logic to more closely follow the spec.

FabianHenneke commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <syslog.h>
+#include <unistd.h>
```
I hope that I got it down as far as possible.

FabianHenneke commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+
+int main(int argc, char **argv) {
+        _cleanup_(udev_unrefp) struct udev *udev = NULL;
+        _cleanup_(udev_device_unrefp) struct udev_device *dev = NULL;
```
Thank you for the advice, I wasn't even aware of the deprecation. I'm entirely new to systemd, so please doublecheck my use of the API and let me know if something else is not following best practices.

FabianHenneke commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,114 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
Since I based this on an existing GPL project, I decided to keep the modification information in. Let me know if this can also go.

poettering replied:
our code is generally LGPL2, not GPL. it's OK to copy in GPL code if it's at the fringes, like it is here. But I think it would be a good idea to add an explicit line that clarifies why this file is GPL and not LGPL.

FabianHenneke replied:
At the current state, this commit doesn't really share a line with the original code anymore. Maybe this would even warrant removing the "modified" header and licensing under LGPL?

FabianHenneke replied:
I went ahead and changed the license to LGPL. I hope @amluto approves, but I can always revert this change if it should not be appropriate.

poettering replied:
Maybe add a comment "Inspired by …" to be friendly

FabianHenneke commented on src/udev/hidraw_id/hidraw_id_desc.h:
```diff
@@ -0,0 +1,14 @@
+struct HidrawIdDescResult {
+        bool is_security_token: 1;
+        bool is_u2f_token: 1;
+};
```
I'm sorry, I don't fully understand what you would want me to change here. Since something can be both a security token (general term) and a U2F token (a specific instance of a security token, although the only existing one so far...), I am not sure this struct can be cleanly mapped onto an enum. I could turn these flags into int-valued enums and or them together if this is the preferred style in systemd.

yuwata replied:
Not sure, but I thought something the following:
```
typedef enum HidrawIdTokenType {
    HIDRAW_ID_SECURITY_TOKEN = 1 << 0,
    HIDRAW_ID_U2F_TOKEN      = 1 << 1,
} HidrawIdTokenType;
```
and then for example,
```
*ret = HIDRAW_ID_SECURITY_TOKEN | HIDRAW_ID_U2F_TOKEN;
```

(btw, how about to use `typedef` to drop `struct`.)

poettering replied:
hmm, so the current code only sets both bools or neither, never one without the other. So why have this at all?

FabianHenneke replied:
The reason for that is that `ID_U2F_TOKEN` is set by the original `u2f_hidraw_policy` created by @amluto and for example packaged into Fedora. Since there may already be userspace software that depends on `ID_U2F_TOKEN` instead of the more general `ID_SECURITY_TOKEN`, I decided to keep this in for the sake of backwards compatibility. Would you suggest to remove it?

FabianHenneke replied:
@yuwata I generally prefer structs over enum flags since they are typesafe and easier to query, but can always adapt if you prefer it the other way. I will definitely use `typedef`, thanks for the suggestion.

poettering replied:
hmm, so these aren#t orthogonal bits, no?

I mean, the current code only sets both or neither, so this doesn't have to be a bitfield or struct at all, but a plain boolean. But even if this code could detect two kinds of tokens: U2F and plain ones, then it every U2F one is a also a security token, right, hence this should just be a plain enum in that case, not a bitfield.

But what am I missing?

FabianHenneke replied:
Yes, the current struct members are not orthogonal, one always implies the other. I originally thought that the struct would make it easier for the caller to take action since they wouldn't have to know about these implications, but agree that this may just be complicating things. I will convert to an enum and we can see whether this is satisfactory.

amluto replied:
> The reason for that is that `ID_U2F_TOKEN` is set by the original `u2f_hidraw_policy` created by @amluto and for example packaged into Fedora. Since there may already be userspace software that depends on `ID_U2F_TOKEN` instead of the more general `ID_SECURITY_TOKEN`, I decided to keep this in for the sake of backwards compatibility. Would you suggest to remove it?

As far as I know, ID_U2F_TOKEN is only used by u2f_hidraw_policy's scripts -- no one has ever told me they use it.  I plan to deprecate ID_U2F_POLICY once this new code lands in systemd and makes it into a release.

FabianHenneke commented:
Thank you for the very detailed comments! I hope that I addressed all of them.
The UB detected by the fuzzer were caused by implicit conversions to (signed) int that I missed. As a consequence, I have added explicit casts to unsigned types to any bitwise operation.
I have also rewritten the parser logic to align more with the way usages actually work per spec - they are always 32 bit, but the upper 16 bit can be set via a usage page.

yuwata commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,104 @@
+        fd = open(desc_path, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+        if (fd < 0)
+                return log_device_error_errno(
+                        hid_device, errno, "Failed to open report descriptor at '%s': %m", desc_path);
```
We usually use the following indentation when line wraps.
```
return log_device_error_errno(hid_device, errno, 
                              "Failed to open report descriptor at '%s': %m", desc_path);
```

poettering commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,104 @@
+                return EXIT_FAILURE;
+        }
+
+        if (streq(argv[1], "--rule")) {
```
hmm, why this?

FabianHenneke replied:
The `if` is there so that the program can be run outside the context of a udev rule (e.g. troubleshooting). I renamed the command line flag to `--rule` to indicate that it is supposed to be used within a rule, as I wasn't sure whether `udev` is a good description for its context. Would you say that `--udev` or something else entirely is better?

yuwata replied:
How about the following:
`argc == 1`: use `environ`, 
`argc == 2`: use `argv[1]`.

FabianHenneke replied:
I like this idea a lot and just implemented it.

poettering commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,102 @@
+                                usage = value;
+                        else
+                                usage = (uint32_t) ((uint32_t) usage & 0xffff0000) |
+                                        ((uint32_t) value & 0x0000ffff);
```
unless I am mistaken `usage` might be used un-initialized here

FabianHenneke replied:
Thanks, completely missed the missing initializer. The fuzzer caught it as well @evverx, I'm now fixing and testing it.

evverx commented:
The fuzzer seems to be failing under MSan:
```sh
==3129==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x4d0cc1 in parse_hid_report_desc /home/vagrant/systemd/build/../src/udev/hidraw_id/hidraw_id_desc.c:93:29
    #1 0x4d0044 in LLVMFuzzerTestOneInput /home/vagrant/systemd/build/../src/fuzz/fuzz-hidraw-id-desc.c:23:13
    #2 0x4359dd in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x4359dd)
    #3 0x436965 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool*) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x436965)
    #4 0x438538 in fuzzer::Fuzzer::MutateAndTestOne() (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x438538)
    #5 0x43b18f in fuzzer::Fuzzer::Loop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, fuzzer::fuzzer_allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x43b18f)
    #6 0x42acd7 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x42acd7)
    #7 0x41e936 in main (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x41e936)
    #8 0x7f096422a412 in __libc_start_main (/lib64/libc.so.6+0x24412)
    #9 0x41e96d in _start (/home/vagrant/systemd/out/fuzz-hidraw-id-desc+0x41e96d)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/vagrant/systemd/build/../src/udev/hidraw_id/hidraw_id_desc.c:93:29 in parse_hid_report_desc
  ORIGIN: invalid (0). Might be a bug in MemorySanitizer origin tracking.
    This could still be a bug in your code, too!
Exiting
MS: 4 CrossOver-ChangeByte-ChangeBit-ChangeBit-; base unit: 86516c6a0099837ca3dcd35174c141c78b8689d5
0x34,0x14,0x9,0x1,
4\x14\x09\x01
```
The previous version didn't crash under MSan as far as I can remember.

FabianHenneke commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,102 @@
+                /* The first byte of a short item encodes tag, type and size. */
+                tag = (uint8_t) desc[pos] >> 4; /* Bits 7 to 4 */
+                type = (uint8_t) ((uint8_t) desc[pos] >> 2) & 0x3; /* Bits 3 and 2 */
+                size_code = (uint8_t) desc[pos] & 0x3; /* Bits 1 and 0 */
```
I agree that these ones are, but the ones further down are not, even though they are superficially very similar. My general policy now that I touch the original parser code would be to add casts after any bitwise operation on unsigned types, in order to prevent the hard-to-spot bugs that come with them (like the one that was found by UBsan in an earlier version of this PR). But if you prefer I could think more about which of these casts are essential and which ones aren't, and remove some of them.

yuwata commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,94 @@
+                        return -EINVAL;
+                value = 0;
+                for (size_t i = 0; i < size; i++)
+                        value |= (uint32_t) desc[pos + i] << (8 * i);
```
I think only this cast is necessary.

FabianHenneke replied:
I removed all casts that either applied to unsigned types already of the same size as an `int` or to bitwise operations other than up-shifts, since these are somewhat clearly okay. This instance is actually the only that remains, which is convenient.

yuwata commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,94 @@
+                        /* Detect U2FHID/CTAPHID security tokens based on their full usage. */
+                        if (usage == FIDO_FULL_USAGE_CTAPHID) {
+                                local_ret = HIDRAW_ID_U2F_TOKEN;
+                        }
```
Please drop unnecessary `{}`.

poettering commented on rules/60-hidraw_id.rules:
```diff
@@ -0,0 +1,7 @@
+# do not edit this file, it will be overwritten on update
```
the Fixes: should be the last line in the commit msg, not in the title really

yuwata commented on src/udev/hidraw_id/hidraw_id_desc.h:
```diff
@@ -0,0 +1,14 @@
+
+typedef enum HidrawIdDescResult {
+        HIDRAW_ID_SECURITY_TOKEN = 1 << 0,
+        HIDRAW_ID_U2F_TOKEN      = 1 << 1,
```
As @poettering commented, please make this simple enum, that is please drop `= 1 << x`.

poettering replied:
or drop it altogether, the current code will only ever set HIDRAW_ID_U2F_TOKEN anyway, no? why have the other bit?

evverx commented:
I've just created a couple of fuzz targets on Fuzzit and uploaded a corpus I generated by brute force :-) Travis CI should be green now.

poettering commented:
So I have the suspicion we should have both ID_SECURITY_TOKEN and ID_U2F_TOKEN in the udev database, because there are tokens that support U2F and tokens that do not. However I am pretty sure the tool this PR includes should only check for the latter and then set both vars at once. i.e. ID_SECURITY_TOKEN should be set on all devices that are security tokens including pre-u2f yubikeys and such, but ID_U2F_TOKEN only on the U2F capable ones.

Ideally we'd set ID_SECURITY_TOKEN for all those non-U2F security tokens too, but I doubt we can recognize them generically, but would have to ship a hwdb file for that, dunno. not sure it's worth it though given we don't know if anyone actually ever made use of the device nodes for non-U2F devices anyway...

amluto commented:
A fundamental problem with “U2F” is that there is FIDO1 and FIDO2.  And there are devices that speak U2F that aren’t HID, e.g. U2F over NFC.

So if systemd wants to distinguish by protocol, some better strings might make sense.

FabianHenneke commented:
Based on my understanding, at least for the purpose of HID device discovery, FIDO1 and FIDO2 (or rather CTAP1 and CTAP2) are not distinguishable. Both use the same report descriptor and can only be distinguished by sending CTAP2 commands to the token. From this point of view, the correct string more specific than `ID_SECURITY_TOKEN` would probably be something like `ID_CTAPHID_TOKEN`. In the way they are designed, the transport layers NFC and BLE of U2F/FIDO2 should never play a role for systemd/udev.

There are relevant cases of non-U2F/CTAP security tokens. For example, YubiKeys offer a configuration interface + OTP and static password functionality over HID. But features like this are probably very hard to detect generically.

@poettering If you think that it would make more sense, I could "rebrand" the entire CL as something more specific than "hidraw_id" and thus make it clear that this program will always be about security tokens (maybe even CTAP-compatible ones) only.

nickray commented:
Just as a comment from the sidelines: There was never a U2F standard, only a proposal, which is now subsumed under CTAP1 (vs. CTAP2 with extended capabilities used in FIDO2). So it would be good not to reference "U2F" in the string. In practice, CTAP2 keys also implement CTAP1/U2F, and I don't think pre-U2F keys merit mention.
Based on the HID usage page's name (https://www.usb.org/sites/default/files/hutrr48.pdf), `ID_FIDO_TOKEN` (or possibly `ID_CTAPHID_TOKEN`) would to make the most sense, and cover both older U2F and newer FIDO2 keys.
What are `ID_SECURITY_TOKEN`s supposed to be?

FabianHenneke commented:
I updated the commit to address most of the remaining comments.

Based on our discussion so far, it sounds reasonable to restrict this tools scope to detecting FIDO CTAP1 (also called "U2F") and CTAP2 HID devices. I created a new version of the current commit on a new branch of my fork: https://github.com/FabianHenneke/systemd/commit/2b96a8dea704d4223b25027da1e63b2c7ff35cf7. In this commit, I renamed everything to `fido_id` and simplified the logic based on the reduced scope. Let me know if you prefer this version and I will replace the current one.

poettering commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,93 @@
+                        if (size == 4)
+                                usage = value;
+                        else
+                                usage = (usage & 0xffff0000) | (value & 0x0000ffff);
```
hmm, on most archs int is 32bit, hence 0xffff0000 is signed though you apply it to an unsigned value `usage`. I think it would be nice to suffix this constant with "U" to avoid the weirdness of writing a signed value as unsigned but then applying it as unsigned, too

FabianHenneke replied:
I added explicit `u`s to all three unsigned constants, matching the `#define`d constants. I can convert to capital `U` if that is more in line with the rest of the code.

poettering commented:
lgtm besides the signed hex constant thing. I figure calling this fido_id is nicer though, @keszybz @yuwata opinions on that?

I would probably keep two distinct properties in place for this. The way I see it ID_SECURITY_TOKEN just marks any security tokens, including first generation yubikeys and such. It is useful for tools that need to wait until a token is inserted or even for UIs in case they want to filter input masks for such devices if they are the kind that generates keypress events for OTP stuff (and shouldn't be allowed to generate C-A-D for example). Hence: let's leave ID_SECURITY_TOKEN very generic, very vague, applying to all kinds of devices and then may add ID_FIDO_TOKEN for the specific types that support FIDO. Does that make sense?

poettering commented:
(and yes I much prefer the is_fido_security_token_desc() prototype that just returns 0, 1 or negative error like in the other version)

keszybz commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,100 @@
+        log_open();
+
+        if (argc > 2) {
+                fprintf(stderr, "Usage: hidraw_id [SYSFS_PATH]\n");
```
```C
return log_error_errno(SYNTHETHIC_ERRNO(EINVAL), "Usage: hidraw_id [SYSFS_PATH]");
```

keszybz commented on src/udev/hidraw_id/hidraw_id.c:
```diff
@@ -0,0 +1,100 @@
+        fd = open(desc_path, O_RDONLY | O_NOFOLLOW | O_CLOEXEC);
+        if (fd < 0)
+                return log_device_error_errno(hid_device, errno,
+                        "Failed to open report descriptor at '%s': %m", desc_path);
```
I'd expect this to be aligned to `(`. The same below.

keszybz commented on src/udev/hidraw_id/hidraw_id_desc.c:
```diff
@@ -0,0 +1,93 @@
+                /* The first byte of a short item encodes tag, type and size. */
+                tag = desc[pos] >> 4; /* Bits 7 to 4 */
+                type = (desc[pos] >> 2) & 0x3; /* Bits 3 and 2 */
+                size_code = desc[pos] & 0x3; /* Bits 1 and 0 */
```
Maybe align the comments to the same column? This will be easier to scan then.

FabianHenneke commented:
Now that @yuwata and @keszybz have voiced their opinions, I went ahead and replaced the PR commit with the branch where everything has been rebranded as/restricted to FIDO.

@evverx Could you adapt the fuzzer targets on the CI system?

evverx commented:
@FabianHenneke sure. I renamed the fuzz targets on Fuzzit and restarted the last job on Travis CI.

poettering commented:
lgtm, but let's merge this after the v243 release

yuwata commented:
LGTM too.

nickray commented:
Woo! Thanks everyone involved!

FabianHenneke commented:
Thanks for the thorough review and timely merge!

poettering commented:
@FabianHenneke hmm, so i tried this with my two Yubikeys (4 and 5c), and in neither of the cases a fido device was detected... the report descriptor is read and then is_fido_security_token() returns 0... any idea what's going on?

which type of device did you test this with?

the usb data reports both as "Yubico YubiKey OTP+FIDO+CCID" so they definitely support FIDO, yet are not recognized by the tool?

poettering commented:
(ah, ignore me, it exposes two hidraw devices, it works for the other one... christ)

fbruetting commented:
Would you like to cherry-pick this for the v239 series, so that it could land in RHEL/CentOS 8/Stream?

keszybz commented:
This is a new feature. It is separate program, so there is no risk of introducing errors to existing code, but the udev rule is included by default, and the effect of this could be quite visible to users. Normally this is not something we would backport. If anything, individual distributions would cherry pick the patch on their own if they wish.
